GVFS/GVFS.Common/Git/GitAuthentication.cs

```diff
@@ -221,6 +255,7 @@ private bool TryCallGitCredential(ITracer tracer, out string errorMessage)
             if (!string.IsNullOrEmpty(gitUsername) && !string.IsNullOrEmpty(gitPassword))
             {
                 this.cachedAuthString = Convert.ToBase64String(Encoding.ASCII.GetBytes(gitUsername + ":" + gitPassword));
+                this.storedCachedAuthString = false;

```

Comment:
I believe there is no guarantee, the credentials here worked.  This is just what GCM returned.

Reply:
Jessica is correct. We don't know if the credentials worked or not, just that we got some credentials back from Git/GCM. The user could have for example signed into the wrong account, which would return 'valid' credentials but not valid for accessing the particular Azure Repos repository/endpoint.

The `GitAuthentication` type here is actually following the same model as `git-credential` in that there is:
1. a way to get credentials `fill`/`TryGetCredentials(...)`,
2. a way to indicate credentials did _not_ work `reject`/`Revoke(...)`, and
3. a way to indicate credentials _did_ work `approve`/`ConfirmCredentialsWorked(...)`.

The `HttpRequestor` type (which is the only product-code caller of `GitAuthentication.TryGetCredentials(...)`) already calls all of the above methods, it's just that `GitAuthentication` itself never called through to `git-credential approve`.

Reply:
We actually had a bit of debate here regarding, so pulling in @jamill for his feedback.

Am I correct in saying the new GCM will return whatever is stored if the string matches?  Aka no validation on GCM?  For *new* keys we would have had to validate we were able to get the PAT against the org.

Reply:
New GCM does zero validation of what is stored in the credential store/Keychain. We hand whatever we have straight back for the caller to determine if it's correct and valid. That's what the `approve` and `reject` actions are for on `git-credential`.

Encoding what is considered a 'valid' credential isn't as clear cut as you'd think and depends on what you're doing to do with it, hence new GCM doesn't. Windows GCM does do some  call to see if the repository 'upload_pack' endpoint can be reached.

If the credential is bad (say it has expired) then doing a `git fetch` will fail the first time, and the bad credential will be deleted. A subsequent call will result in a prompt (since there is now no credential stored, good or bad):

```
$ git fetch
fatal: Authentication failed for '<URL>'
$ git fetch
# prompt is shown & fetch is executed successfully
```

If we want to get into the business of verifying credentials before returning to Git then we should be very careful about thing. We can end up getting into states were we're being too smart, for example for NTLM/Kerberos or other federated tokens via ADAL where in Windows GCM we're saying the credentials are 'valid' for one purpose then then Git finds out they're not and we start ignoring requests to delete 'invalid' credentials.

Not to mention we should then be thinking about what constitutes 'valid' for VFS4Git vs normal Git operations, and other host providers like GitHub or on-prem TFS (and which versions?).